projs/shadow/shadow-engine/core/src/profile/Profiler.cpp

```diff
@@ -0,0 +1,556 @@
+#include <shadow/profile/Profiler.h>
+#include <vector>
+#include "shadow/assets/fs/iostream.h"
+#include "shadow/util/Syncronization.h"
+#include "shadow/core/Time.h"
+#include <shadow/core/Thread.h>
+#include <shadow/platform/Common.h>
+
+#ifdef _WIN32
+#define INITGUID
+#define NOGDI
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <evntcons.h>
+#include <atomic>
+#include <cstring>
+
+#endif
+
+namespace SH {
+
+  // Allows the Profiler to be used on multiple threads simultaneously, separating the data collected.
+  // Every thread has its own set of open blocks, its own mutex, and its own visibility toggles.
+  // Consider the ThreadContext as a horizontal bar on a profiler menu, showing what one thread is doing at any given time.
+  struct ThreadContext {
+
+    ThreadContext(size_t bufferSize) : oms(), openBlocks() {
+        oms.resize(bufferSize);
+        openBlocks.reserve(64);
+    }
+
+    std::vector<uint32_t> openBlocks;
+    // Used as the data buffer.
+    ShadowEngine::OutputMemoryStream oms;
+    size_t begin = 0;
+    size_t end = 0;
+    ShadowEngine::Mutex mut;
+    std::string name;
+    bool show = false;
+    size_t threadID;
+  };
+
+  // This ifdef contains a bunch of the infrastructure necessary for performance monitoring under Windows.
```

Comment:
This section should be extracted to the platforms folder so I can even try to add Linux support and later of course for the console support

Reply:
Fair point, though if i do that i'll have to duplicate the entire code between Linux and Windows, because these are static internal classes, not exposed in the header.

It's here for the DRY principle. A lot would be repeated if these were in their own file.